fergus_emi_ea-2021-04-ndvi-automation.py

```diff
@@ -0,0 +1,758 @@
+def open_clean_bands(band_path,
+                     crop_extent,
+                     valid_range=None,):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_path : string
+        A path to the array to be opened
+    valid_range : tuple (optional)
+        A tuple of min and max range of values for the data. Default = None
+
+
+    Returns
+    -----------
+    arr : xarray DataArray
+        An xarray DataArray with values that should be masked 
+        set to 1 for True (Boolean)
+    """
+    # TODO add tests to ensure the arrays are the same .shape
+    band = rxr.open_rasterio(band_path, masked=True).rio.clip(crop_extent.geometry,
+                                                              from_disk=True).squeeze()
+
+    # Only run this step if a valid range tuple is provided
+    if valid_range:
+        mask = ((band < valid_range[0]) | (band > valid_range[1]))
+        band = band.where(~xr.where(mask, True, False))
+
+    return band
+
+
+###############
+#  FUNCTION TO OPEN and CLEAN UP SPECIFIC LANDSAT BANDS 
+#  AND MASK USING PIXEL QA LAYER (CLOUD COVER) AND CALCULATE NDVI
+def mask_crop_ndvi(band_paths,
+                   crop_bound,
+                   pixel_qa_path,
+                   vals):
+    """Open and mask a single landsat band using the open_clean_bands function, 
+    calculate NDVI and mask using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_paths : string
+        A path to the landsat scene of interest
+    all_bands : list
+        A list containing two xarray objects for landsat bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to crop the raster data using rasterio mask().
+    pixel_qa_path: string
+        A path to a pixel qa tif file.
+    vals: list
+        A list of values needed to create the cloud mask
+
+
+    Returns
+    -----------
+    ndvi_crop : Xarray Dataset
+        a cropped and masked xarray object containing NDVI values
+    """
+    # For loop using open_clean_bands function
+    all_bands = []
+    for aband in band_paths:
+        cleaned_band = open_clean_bands(band_path=aband,
+                                        crop_extent=crop_bound,
+                                        valid_range=(0, 10000))
+        all_bands.append(cleaned_band)
+
+    crop_json = crop_bound.geometry
+
+    # Open and clip qa layer
+    pixel_qa = rxr.open_rasterio(pixel_qa_path[0], masked=True).rio.clip(crop_json,
+                                                                         from_disk=True).squeeze()
+
+    # Calculate NDVI
+    ndvi_xr = (all_bands[1]-all_bands[0]) / (all_bands[1]+all_bands[0])
+
+    # Apply cloud mask to NDVI
+    ndvi_mask = ndvi_xr.where(~pixel_qa.isin(vals))
+
+    return ndvi_mask
+
+
+###############
+#  FUNCTION to create a pandas dataframe from NDVI and grab site and date from file path
```

Comment:
This function could be placed at the top of the notebook with your other functions